build.go

```diff
@@ -0,0 +1,355 @@
+
+	composerGlobalBuffer := bytes.NewBuffer(nil)
+
+	execution := pexec.Execution{
+		Args: append([]string{"global", "require", "--no-progress"}, globalPackages...),
+		Dir:  composerGlobalLayer.Path,
+		Env: append(os.Environ(),
+			"COMPOSER_NO_INTERACTION=1", // https://getcomposer.org/doc/03-cli.md#composer-no-interaction
+			fmt.Sprintf("COMPOSER_HOME=%s", composerGlobalLayer.Path),
+			fmt.Sprintf("PHPRC=%s", composerPhpIniPath),
+			"COMPOSER_VENDOR_DIR=vendor", // ensure default in the layer
+			fmt.Sprintf("PATH=%s", path),
+		),
+		Stdout: composerGlobalBuffer,
+		Stderr: composerGlobalBuffer,
+	}
+	err = composerGlobalExec.Execute(execution)
+
+	if err != nil {
+		logger.Subprocess(composerGlobalBuffer.String())
+		return "", err
+	}
+
+	composerGlobalBin := filepath.Join(composerGlobalLayer.Path, "vendor", "bin")
+
+	if os.Getenv(BpLogLevel) == "DEBUG" {
+		logger.Debug.Subprocess(composerGlobalBuffer.String())
+		logger.Debug.Action("Adding global Composer packages to PATH:")
+		files, err := ioutil.ReadDir(composerGlobalBin)
+		if err != nil {
+			return "", err
+		}
+		for _, f := range files {
+			logger.Debug.Detail(fmt.Sprintf("- %s", f.Name()))
+		}
+	}
+
+	return composerGlobalBin, nil
+}
+
+func runComposerInstall(
+	logger scribe.Emitter,
+	context packit.BuildContext,
+	composerInstallOptions DetermineComposerInstallOptions,
+	composerPhpIniPath string,
+	path string,
+	composerInstallExec Executable,
+	workspaceVendorDir string,
+	calculator Calculator) (packit.Layer, string, error) {
+
+	launch, build := draft.NewPlanner().MergeLayerTypes(ComposerPackagesDependency, context.Plan.Entries)
+
+	composerPackagesLayer, err := context.Layers.Get(ComposerPackagesLayerName)
+	if err != nil {
+		return packit.Layer{}, "", err
+	}
+
+	composerJsonPath, composerLockPath, _, _ := FindComposerFiles(context.WorkingDir)
+
+	layerVendorDir := filepath.Join(composerPackagesLayer.Path, "vendor")
+
+	composerLockChecksum, err := calculator.Sum(composerLockPath)
+	if err != nil {
+		return packit.Layer{}, "", err
+	}
+
+	logger.Debug.Subprocess("Calculated checksum of %s for composer.lock", composerLockChecksum)
+
+	if cachedSHA, ok := composerPackagesLayer.Metadata["composer-lock-sha"].(string); ok && cachedSHA == composerLockChecksum {
+		logger.Process("Reusing cached layer %s", composerPackagesLayer.Path)
+		logger.Break()
+
+		composerPackagesLayer.Launch, composerPackagesLayer.Build, composerPackagesLayer.Cache = launch, build, launch
+
+		logger.Debug.Subprocess("Setting cached layer types: launch=[%t], build=[%t], cache=[%t]",
+			composerPackagesLayer.Launch,
+			composerPackagesLayer.Build,
+			composerPackagesLayer.Cache)
+
+		return composerPackagesLayer, layerVendorDir, nil
+	}
+
+	composerPackagesLayer, err = composerPackagesLayer.Reset()
+	if err != nil {
+		return packit.Layer{}, "", err
+	}
+
+	composerPackagesLayer.Launch, composerPackagesLayer.Build, composerPackagesLayer.Cache = launch, build, launch
```

Comment:
Why is `composerPackagesLayer.Cache` set to `launch` instead of `build`?

Reply:
There's an expectation that `composerPackagesLayer` will be contributed only at `Launch` time, and hardly ever at `Build` time.  I may be missing a nuance of how layers are cached, but it felt that if we set `cache = build` then it would effectively never be cached.

`php-start`: https://github.com/paketo-buildpacks/php-start/blob/b4e497fef5b5bad689b2e9992af9a45e33f54587/detect.go#L65-L70

`php-builtin-server`: https://github.com/paketo-buildpacks/php-builtin-server/pull/29

Reply:
I thought I had an example where `cache` was set to the value of `launch`, but I can't find it. Looks like both `npm-install` and `yarn-install` force `cache = true`. Perhaps we should do that here?

https://github.com/paketo-buildpacks/npm-install/blob/main/build.go

https://github.com/paketo-buildpacks/yarn-install/blob/main/build.go

Reply:
@joshuatcasey We never need to set the cache layer to the value of launch. It's confusing but when `launch=true`, the layer metadata will always be restored, the buildpack reads the metadata and leaves it in place, telling the lifecycle to reuse the layer. This is all to say that we don't need to set cache to launch, even though in most cases another buildpack will pull it in at launch (instead of build)

We also don't need to always set `cache=true`, in NPM Install, that's being done for an intermediate build cache (not required by another buildpack), not the final module layer. That's not synonymous with what we need here. In Yarn Install, `cache=true` is being hardcode, but looking at the code, it happens [within a codeblock](https://github.com/paketo-buildpacks/yarn-install/blob/217a13c5da3fb28b8c140841540a1db9af1a7373/build.go#L87) that only applies to `build=true`.

Reply:
fixed in latest commit :D